### ISSUE TITLE
Faraday::Adapter::Test can't stub by URL, only by path

### ISSUE THREAD

**Issue by Peeja** [lostisland/faraday#358]
I'm trying to write a spec for an object which will fetch content from a URL. (I'm abstracting away lots of domain-specific stuff that's not relevant to this issue here.) I want to stub the Faraday request that fetches the content. I want the spec to be something like so:

``` ruby
let(:http_connection) {
  Faraday.new do |builder|
    builder.adapter :test, http_stubs
  end
}

let(:http_stubs) { Faraday::Adapter::Test::Stubs.new }
let(:fetcher) { ThingFetcher.new(http_connection: http_connection) }

it "gets the right thing" do
  http_stubs.get("http://example.test/things/1") { [200, {}, 'a thing'] }

  expect {
    fetcher.fetch("http://example.test/things/1")
  }.to change { Thing.count }.by(1)

  expect(Thing.last.contents).to be('a thing')
end
```

But it turns out that doesn't work because

```
Faraday::Adapter::Test::Stubs::NotFound: no stubbed request for get /things/1
```

[`Faraday::Adapter::Test` normalizes the URL, stripping off the host information](https://github.com/lostisland/faraday/blob/3dc615a72eae5cec55d2c73dfdd8efea00f2c283/lib/faraday/adapter/test.rb#L146) before it looks it up in the stubs. But I _want_ to assert that it uses the right host. I want to stub the entire URL.

How can I do that? And if I can't, is there a path toward making it so one can?

**Comment by erik-escobedo** [comment]
What the `any_volunteer?` tag means? I do wanna work in this and have an idea about how to fix it. Should I start working on the PR? 😮

**Comment by iMacTia** [comment]
Hi @erik-escobedo, yes the `any_volunteer?` tag was introduced before the new official `help wanted` and `good first issue`. I suppose it would be better to update them. Anyway, I'm happy if you want to help on this one. I would suggest you expose your idea before working and opening the PR so we can discuss about it beforehand (and hopefully save some of your time)

**Comment by erik-escobedo** [comment]
Well, as @Peeja pointed out, when the URL is provided the `host` is striped off by `Faraday::Utils.normalize_path(env[:url])` therefore, the `matched?` method completely ignores it. My plan is to store the host in a new variable and make sure the `host` provided on the stub matches the one provided on the `request`. If no host is provided, the variable stores `nil` and nothing changes. Even tests won't need to be updated. Then, only 2 new tests need to be written:

1. For when host is provided and matches
2. For when host is provided and doesn't match

**Comment by iMacTia** [comment]
That sounds good to me.
As long as the host is optional, old tests don't need to be changed and new tests are added, I think we're on the right track 👍

**Comment by erik-escobedo** [comment]
Awesome. I'll work on it ASAP 😃

**Comment by iMacTia** [comment]
Closed with #741
